### PR TITLE
Add AgentMarket.get_token_balance, with support for Omen

### PR DESCRIPTION
**Pull request by evangriffiths**
Carrying on with supporting a 1-to-1 mapping between AgentMarket functions and functions required by a 'general agent' (e.g. microchain agent in PMA).

Support for manifold and polymarket to come later

### CODE REVIEW

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 1]
```diff
@@ -206,6 +206,23 @@ def get_contract(self) -> OmenFixedProductMarketMakerContract:
...
+        # Find position matching market_outcome
+        index_set = self.get_index_set(outcome)
+        position_for_index_set = next(p for p in positions if index_set in p.indexSets)
+        position_as_int = int(position_for_index_set.id.hex(), 16)
+
+        balance = OmenConditionalTokenContract().balanceOf(
+            from_address=Web3.to_checksum_address(user_id),
+            position_id=position_as_int,
+        )
+        return Decimal(balance)
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply prediction_market_agent_tooling/markets/omen/omen.py:224]: Makes sense coderabbitai! We should check for cases where outcome is not contained in `market.outcomes` and return a 0 balance.

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 2]
```diff
@@ -206,6 +206,23 @@ def get_contract(self) -> OmenFixedProductMarketMakerContract:
...
+    def get_index_set(self, outcome: str) -> int:
+        return self.get_outcome_index(outcome) + 1
+
+    def get_token_balance(self, user_id: str, outcome: str) -> Decimal:
+        positions = OmenSubgraphHandler().get_positions(self.condition.id)
+
+        # Find position matching market_outcome
+        index_set = self.get_index_set(outcome)
+        position_for_index_set = next(p for p in positions if index_set in p.indexSets)
+        position_as_int = int(position_for_index_set.id.hex(), 16)
```
Comment (kongzii) [inline prediction_market_agent_tooling/markets/omen/omen.py:218]: This seems like a common thing we are doing, wdyt about adding method `as_int` to https://github.com/gnosis/prediction-market-agent-tooling/blob/b081d0a3d764e22c5892d952ec8cf7f862f8f38f/prediction_market_agent_tooling/tools/hexbytes_custom.py#L42?

Comment (kongzii) [reply prediction_market_agent_tooling/markets/omen/omen.py:218]: Btw there is this function from @gabrielfior  already: https://github.com/gnosis/prediction-market-agent-tooling/blob/877f281b60084c439ef0c96b009317f4a2b209b0/prediction_market_agent_tooling/markets/omen/omen.py#L537

Seems like something like this should work:

```python
index_set = self.get_index_set(outcome)
balances = get_conditional_tokens_balance_for_market(self, Web3.to_checksum_address(user_id))
return balances[index_set]
```

File: prediction_market_agent_tooling/markets/omen/omen.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kongzii) [reply prediction_market_agent_tooling/markets/omen/omen.py:224]: > Makes sense. We should check for cases where outcome is not contained in market.outcomes and return a 0 balance.

Returning 0 should mean that the outcome does exist, but the user doesn't have any tokens for it; asking for an outcome that's not in the market feels like a bug.

So I'm good with Evan's current implementation (except for the suggestion in another comment to reuse the existing function). That's doing

```
def get_index_set(self, outcome: str) -> int:
        return self.get_outcome_index(outcome) + 1

def get_outcome_index(self, outcome: str) -> int:
        try:
            return self.outcomes.index(outcome)
        except ValueError:
            raise ValueError(f"Outcome `{outcome}` not found in `{self.outcomes}`.")
```

and so will throw a ValueError if wrong outcome is passed in.

File: prediction_market_agent_tooling/markets/omen/omen.py (hunk 2 above)
Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: 🤩 Oo that's handy. Great suggestion!

Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: > This seems like a common thing we are doing, wdyt about adding method `as_int`

Also a good idea. Will leave to the next person to add another case of it 😆

Comment (kongzii) [reply prediction_market_agent_tooling/markets/omen/omen.py]: We can just do this, look:

@coderabbitai Please create an issue with the name "Add method `as_int` to the HexBytes class", thanks!

Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: oo cool feature

Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: would be even better if it linked to this conversation in the issue description

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 3]
```diff
@@ -206,6 +206,16 @@ def get_contract(self) -> OmenFixedProductMarketMakerContract:
...
 
+    def get_index_set(self, outcome: str) -> int:
+        return self.get_outcome_index(outcome) + 1
+
+    def get_token_balance(self, user_id: str, outcome: str) -> Decimal:
+        index_set = self.get_index_set(outcome)
+        balances = get_conditional_tokens_balance_for_market(
+            self, Web3.to_checksum_address(user_id)
+        )
+        return Decimal(balances[index_set])
```
Comment (kongzii) [inline prediction_market_agent_tooling/markets/omen/omen.py:217]: Just double check if agent works with Weis and doesn't expect this to be in xDai?

**Review by kongzii — APPROVED** [review]

File: prediction_market_agent_tooling/markets/omen/omen.py (hunk 3 above)
Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py:217]: Hmm, having a bit of a confused moment here. What does a token balance refer to here. Just assume we convert the value to xDai for now. Is it

- "The amount of xDai I bet on that outcome for that market" or
- "The current value of my tokens on that outcome if that market were to resolve in favour of that outcome"
- or something else

?

i.e. if I bet 1 xDai on Yes on a market, then call `get_token_balance` on that market, will it return `1 xDai` or will it depend on the current p_yes of the market?

Comment (kongzii) [reply prediction_market_agent_tooling/markets/omen/omen.py:217]: I think it should be amount of shares that we have. So if we buy 26.79 shares for 10 xDai, get_token_balance should always be 26.79 shares, which may become 26.79 xDai if we were correct.

> The current value

Does the value of the shares that were already bought change over time? 😱 

Screenshot by Dropbox Capture

Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py:217]: Wonder if `get_token_balance`'s return type should be `BetAmount` in that case - i.e. have amount and currency

Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py:217]: Or a `TokenAmount`, and then BetAmount and ProfitAmount 
https://github.com/gnosis/prediction-market-agent-tooling/blob/main/prediction_market_agent_tooling/markets/data_models.py#L21

can just be aliases

Comment (kongzii) [reply prediction_market_agent_tooling/markets/omen/omen.py:217]: > Or a `TokenAmount`, and then BetAmount and ProfitAmount https://github.com/gnosis/prediction-market-agent-tooling/blob/main/prediction_market_agent_tooling/markets/data_models.py#L21
> 
> can just be aliases

Yeah that looks good

**Pull request merged**